api/include/opentelemetry/http/http_client.h

```diff
@@ -0,0 +1,161 @@
+
+  // Sets the request Method
+  virtual void SetMethod(nostd::string_view const &method) = 0;
+
+  // Gets the request URI
+  virtual void SetUrl(nostd::string_view const &url) = 0;
+
+  // Gets the request Headers
+  virtual HttpHeaders *GetHeaders() const = 0;
+
+  // Sets the request body
+  virtual void SetBody(const uint8_t *const body, const size_t len) = 0;
+
+  // Gets the request body
+  virtual void GetBody(uint8_t *body, size_t &len) = 0;
+};
+
+// The HttpResponse class represents a Response object.
+// Individual HTTP client implementations can implement the response object
+// in the most efficient way. Either copy all of the underlying data to a new
+// structure, and provide it to the callback; or keep the real
+// response data around, forward the methods, and retrieve the individual
+// values directly one by one.
+class HttpResponse
+{
+public:
+  // Gets the response ID.
+  virtual const nostd::string_view &GetId() = 0;
+
+  // Get the response result code.
+  virtual HttpResult GetResult() = 0;
+
+  // Gets the response status code.
+  virtual unsigned GetStatusCode() = 0;
+
+  // Gets the response headers.
+  virtual HttpHeaders *GetHeaders() = 0;
+
+  // Gets the response body.
+  virtual void GetBody(uint8_t *body, size_t &len) = 0;
+};
+
+// The HttpResponseCallback class receives HTTP client responses
+class HttpResponseCallback
+{
+public:
+  // Called when an HTTP request completes.
+  // The passed response object contains details about the exact way the
+  // request finished (HTTP status code, headers, content, error codes
+  // etc.). The ownership of the response object is transferred to the
+  // callback object. It can store it for later if necessary. Finally, it
+  // must be deleted using its virtual destructor.
+  virtual void OnHttpResponse(HttpResponse *response) = 0;
+};
+
+// The HttpClient class is the interface for HTTP client implementations.
+class HttpClient
+{
+public:
+  // Creates an empty HTTP request object.
+  // The created request object has only its ID prepopulated. Other fields
+  // must be set by the caller. The request object can then be sent
+  // using SendRequestAsync(). If you are not going to use the request object
+  // then you can delete it safely using its virtual destructor.
+  virtual HttpRequest *CreateRequest() = 0;
+
+  // Begins an HTTP request.
+  // The method takes ownership of the passed request, and can destroy it before
+  // returning to the caller. Do not access the request object in any
+  // way after this invocation, and do not delete it.
+  // The callback object is always called, even if the request is
+  // cancelled, or if an error occurs immediately during sending. In the
+  // latter case, the OnHttpResponse() callback is called before this
+  // method returns. You must keep the callback object alive until its
+  // OnHttpResponse() callback is called.
+  virtual void SendRequestAsync(HttpRequest *request, HttpResponseCallback *callback) = 0;
```

Comment:
I think we may consider dropping the `Async` suffix from the method name. Since we provide a callback argument - perhaps we can just say that the API is async? Same comment applies for `CancelRequestAsync`.. Whereas for cancellation - I think it'd make sense to spell out that the cancellation is synchronous (cause there is practically only one scenario, system / service shutdown, where we are gonna be needing to employ that).

Reply:
Good point. Done the changes.